applications/ShapeOptimizationApplication/custom_utilities/response_functions/face_angle_response_function_utility.cpp

```diff
@@ -0,0 +1,132 @@
+// ==============================================================================
+//  KratosShapeOptimizationApplication
+//
+//  License:         BSD License
+//                   license: ShapeOptimizationApplication/license.txt
+//
+//  Main authors:    Geiser Armin, https://github.com/armingeiser
+//
+// ==============================================================================
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_utilities/response_functions/face_angle_response_function_utility.h"
+#include "utilities/variable_utils.h"
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+FaceAngleResponseFunctionUtility::FaceAngleResponseFunctionUtility(ModelPart& rModelPart, Parameters ResponseSettings)
+	: mrModelPart(rModelPart)
+{
+	const std::size_t domain_size = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+	KRATOS_ERROR_IF(domain_size != 3) << "FaceAngleResponseFunctionUtility can only be used on 3D geometries!" << std::endl;
+
+	mMainDirection = ResponseSettings["main_direction"].GetVector();
+	const double direction_norm = norm_2(mMainDirection);
+	KRATOS_ERROR_IF(domain_size != 3) << "FaceAngleResponseFunctionUtility: 'main_direction' vector norm is 0!" << std::endl;
+	mMainDirection /= direction_norm;
+
+	double min_angle = ResponseSettings["min_angle"].GetDouble();
+	mSinMinAngle = std::sin(min_angle * Globals::Pi / 180);
+
+	std::string gradient_mode = ResponseSettings["gradient_mode"].GetString();
+	if (gradient_mode == "finite_differencing")
+	{
+		double delta = ResponseSettings["step_size"].GetDouble();
+		mDelta = delta;
+	}
+	else
+		KRATOS_ERROR << "Specified gradient_mode '" << gradient_mode << "' not recognized. The only option is: finite_differencing" << std::endl;
+}
+
+double FaceAngleResponseFunctionUtility::CalculateValue()
+{
+	KRATOS_TRY;
+
+	const double value = block_for_each<SumReduction<double>>(mrModelPart.Conditions(), [&](Condition& rCond) {
+		const double g_i = CalculateConditionValue(rCond);
+		if (g_i <= 0) {
+			return 0.0;
+		}
+		return g_i*g_i;
+	});
+
+	mValue = sqrt(value);
+
+	return mValue;
+
+	KRATOS_CATCH("");
+}
+
+void FaceAngleResponseFunctionUtility::CalculateGradient()
+{
+	KRATOS_TRY;
+	// First gradients are initialized
+	VariableUtils().SetHistoricalVariableToZero(SHAPE_SENSITIVITY, mrModelPart.Nodes());
```

Comment:
Probably I am missing something here. 
But, when the `SHAPE_SENSITIVITY` is set to zero here, that means, the sensitivity which is calculated for other objectives before this objective is lost no ? 
This makes it not possible to combine this objective with others ?

Reply:
It is possible, because all responses pass the sensitivity information to the optimizer right after computing them - there they are stored in another generic variable e.g. `DF1DX`.

We are using the `SHAPE_SENSITIVITY` for all responses, just as an intermediate storage. This is not ideal, bot otherwise we would need a dedicated variable for each response.

Reply:
Ok. I see it now !